File: 2015/2/rust/src/main.rs
// use std::fs;
// use std::io::Error;


// fn vec_lines(file_path: &str) -> Result<Vec<String>, Error> {
//     fs::read_to_string(file_path)?
//         .lines()
//         .map(|line| Ok(line.to_string()))
//         .collect()
// }




// fn part_1() -> Result<(), Error>{
//     let data: &Vec<String> = &vec_lines("../input.txt")?;
//     let mut total: i32 = 0;
//     for line in data.iter(){
//         let dimension: Vec<i32> = line
//         .split('x')
//         .filter_map(|s| s.parse::<i32>().ok())
//         .collect();
        
//         let (l,w,h) = (dimension[0], dimension[1], dimension[2]);
//         let lw = l*w;
//         let wh = w*h;
//         let hl = h*l;

//         let smallest = lw.min(wh).min(hl);
//         let area = (2*lw) + (2*wh) + (2*hl) + smallest;
//         total += area;

//     }
//     println!("{total}");
//     Ok(())
// }

// fn part_2() -> Result<(), Error>{
//     let data: &Vec<String> = &vec_lines("../input.txt")?;
//     let mut total: i32 = 0;
//     for line in data.iter(){
//         let mut dimension: Vec<i32> = line
//         .split('x')
//         .filter_map(|s| s.parse::<i32>().ok())
//         .collect();
//         dimension.sort();
//         let warp = (dimension[0] * 2) + (dimension[1] * 2);
//         let bow = dimension[0] * dimension[1] * dimension[2];
//         let ribbon = warp+bow;
//         total += ribbon;

//     }
//     println!("{total}");
//     Ok(())
// }

// fn main(){
//     println!("Part 1");
//     part_1();
//     println!("Part 2");
//     part_2();
//     }

use std::fs;
use std::io::Error;

struct Present {
    length: i32,
    width: i32,
    height: i32,
}

impl Present {
    fn from_str(s: &str) -> Option<Self> {
        let mut dims = s
            .split('x')
            .filter_map(|s| s.parse::<i32>().ok());
        
        match (dims.next(), dims.next(), dims.next()) {
            (Some(l), Some(w), Some(h)) => Some(Present {
                length: l,
                width: w,
                height: h,
            }),
            _ => None,
        }
    }

    fn surface_area(&self) -> i32 {
        let sides = [
            self.length * self.width,
            self.width * self.height,
            self.height * self.length,
        ];
        let smallest = sides.iter().min().unwrap_or(&0);
        2 * sides.iter().sum::<i32>() + smallest
    }

    fn ribbon_length(&self) -> i32 {
        let mut sides = [self.length, self.width, self.height];
        sides.sort_unstable();
        let perimeter = 2 * (sides[0] + sides[1]);
        let volume: i32 = sides.iter().product();
        perimeter + volume
    }
}

fn read_presents(file_path: &str) -> Result<Vec<Present>, Error> {
    fs::read_to_string(file_path)?
        .lines()
        .filter_map(Present::from_str)
        .collect::<Vec<_>>()
        .into_iter()
        .map(Ok)
        .collect()
}

fn calculate_total_paper(presents: &[Present]) -> i32 {
    presents.iter().map(Present::surface_area).sum()
}

fn calculate_total_ribbon(presents: &[Present]) -> i32 {
    presents.iter().map(Present::ribbon_length).sum()
}

fn main() -> Result<(), Error> {
    let presents = read_presents("../input.txt")?;
    
    println!("Part 1");
    println!("{}", calculate_total_paper(&presents));
    
    println!("Part 2");
    println!("{}", calculate_total_ribbon(&presents));
    
    Ok(())
}

File: 2015/2/go/main.go
package main

import (
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

func Part1(file *os.File) (int, error) {
	data, err := io.ReadAll(file)
	if err != nil {
		return 0, fmt.Errorf("Error reading file contents: %w", err)
	}

	total := 0
	lines := strings.Split(strings.TrimSpace(string(data)), "\n")

	for _, line := range lines {
		parts := strings.Split(line, "x")
		if len(parts) != 3 {
			return 0, fmt.Errorf("invalid input format: %s", line)
		}

		l, err := strconv.Atoi(parts[0])
		if err != nil {
			return 0, fmt.Errorf("error converting length: %w", err)
		}

		w, err := strconv.Atoi(parts[1])
		if err != nil {
			return 0, fmt.Errorf("error converting width: %w", err)
		}

		h, err := strconv.Atoi(parts[2])
		if err != nil {
			return 0, fmt.Errorf("error converting height: %w", err)
		}

		side1 := l * w
		side2 := w * h
		side3 := h * l

		smallest := min(side1, min(side2, side3))
		area := (2 * side1) + (2 * side2) + (2 * side3) + smallest

		total += area
	}

	return total, nil
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func Part2(file *os.File) (int, error) {
	data, err := io.ReadAll(file)
	if err != nil {
		return 0, fmt.Errorf("Error reading file contents: %w", err)
	}

	total := 0
	lines := strings.Split(strings.TrimSpace(string(data)), "\n")

	for _, line := range lines {
		parts := strings.Split(line, "x")
		if len(parts) != 3 {
			return 0, fmt.Errorf("invalid input format: %s", line)
		}

		dimensions := make([]int, 3)
		for i, part := range parts {
			dim, err := strconv.Atoi(part)
			if err != nil {
				return 0, fmt.Errorf("error converting dimension: %w", err)
			}
			dimensions[i] = dim
		}

		sort(dimensions)
		wrap := 2*dimensions[0] + 2*dimensions[1]
		bow := dimensions[0] * dimensions[1] * dimensions[2]
		total += wrap + bow
	}

	return total, nil
}

func sort(nums []int) {
	for i := 0; i < len(nums)-1; i++ {
		for j := i + 1; j < len(nums); j++ {
			if nums[i] > nums[j] {
				nums[i], nums[j] = nums[j], nums[i]
			}
		}
	}
}

func main() {
	file, err := os.Open("../input.txt")
	if err != nil {
		fmt.Println("Error opening the file:", err)
		return
	}
	defer file.Close()

	result1, err := Part1(file)
	if err != nil {
		fmt.Println("Error in Part1:", err)
		return
	}

	_, err = file.Seek(0, 0)
	if err != nil {
		fmt.Println("Error resetting file pointer:", err)
		return
	}

	result2, err := Part2(file)
	if err != nil {
		fmt.Println("Error in Part2:", err)
		return
	}

	fmt.Printf("Part 1: %d\n", result1)
	fmt.Printf("Part 2: %d\n", result2)
}
